### PR TITLE
SALTO-7045 Support authenticating NS using OAuth 2.0

### PR DESCRIPTION
**Pull request by dantal4**
Starting from SDF version 2024.2, we cannot longer use token-based authentication in SDF. Instead, there’s a new OAuth authentication for CIs - by generating a certificate, uploading it to NS, and using the certificate Id + private key to authenticate SDF.

There’s no SDF version that supports both methods, so until all users migrated their credentials to OAuth creds, we need to use two versions of the suitecloud package in order to support two formats of credentials - TBA & OAuth.

As for installing the SDF JAR, we should install both versions, while we’re using both versions in the code.

---

_Additional context for reviewer_

---
_Release Notes_: 
Netsuite Adapter:
- Support authenticating NS using OAuth 2.0

---
_User Notifications_: 
None

### CODE REVIEW

**Review by omrilit — COMMENTED** [review]
Amazing work @dantal4 👏

File: packages/netsuite-adapter/e2e_test/credentials_store/adapter.ts [hunk 1]
```diff
@@ -46,14 +58,21 @@ const adapter: Adapter<Args, Credentials> = {
...
   },
-  credentials: async args => ({
-    accountId: toCredentialsAccountId(args.accountId),
-    tokenId: args.tokenId,
-    tokenSecret: args.tokenSecret,
-    suiteAppTokenId: args.suiteAppTokenId,
-    suiteAppTokenSecret: args.suiteAppTokenSecret,
-    suiteAppActivationKey: args.suiteAppActivationKey,
-  }),
+  credentials: async args => {
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/e2e_test/credentials_store/adapter.ts:61]: yes

File: packages/netsuite-adapter/README.md [hunk 2]
```diff
@@ -29,6 +29,21 @@ yarn build
 - Enable SDF in Your NetSuite Account (Admin Only) - follow the instructions under https://<ACCOUNT_ID>.app.netsuite.com/app/help/helpcenter.nl?fid=section_4724921034.html
 - Setup Your Role (prefer Administrator) for SDF Development - follow the instructions under https://<ACCOUNT_ID>.app.netsuite.com/app/help/helpcenter.nl?fid=subsect_1539287603.html
 
+## Authenticate using OAuth
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/README.md:32]: isn't it exactly what it says two lines above it?

File: packages/netsuite-adapter/e2e_test/adapter.test.ts [hunk 3]
```diff
@@ -163,13 +163,18 @@ describe('Netsuite adapter E2E with real account', () => {
...
   jest.setTimeout(1000000)
 
-  each([
-    ['without SuiteApp', false],
-    ['with SuiteApp', true],
-  ]).describe('%s', (_text, withSuiteApp) => {
+  describe.each([
+    ['without SuiteApp', { withSuiteApp: false, withOauth: false }],
+    ['with SuiteApp', { withSuiteApp: true, withOauth: false }],
+    ['with Oauth', { withSuiteApp: false, withOauth: true }],
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/e2e_test/adapter.test.ts:169]: can you explain why?

File: packages/netsuite-adapter/test/client/sdf_client.test.ts [hunk 4]
```diff
@@ -91,7 +91,7 @@ jest.mock('@salto-io/lowerdash', () => ({
 const mockExecuteAction = jest.fn()
 const mockSetCommandTimeout = jest.fn()
 
-jest.mock('@salto-io/suitecloud-cli', () => ({
+jest.mock('@salto-io/suitecloud-cli-legacy', () => ({
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/test/client/sdf_client.test.ts:94]: I thought that adding it to the e2e will cover it, but I can add it there too.

File: packages/netsuite-adapter/src/adapter_creator.ts [hunk 5]
```diff
@@ -171,7 +223,11 @@ export const adapter: Adapter = {
   configType,
   install: async (): Promise<AdapterInstallResult> => {
     try {
-      return await SdkDownloadService.download()
+      const legacyResult = await LegacySdkDownloadService.download()
+      log.info('Installed legacy SDF: %o', legacyResult)
+      const newResult = await NewSdkDownloadService.download()
+      log.info('Installed new SDF: %o', newResult)
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/adapter_creator.ts:229]: it is for simplicity. I'll verify what re-installation is doing.

File: packages/netsuite-adapter/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/README.md:32]: My bad, I thought that the above step is specific to TBH

File: packages/netsuite-adapter/e2e_test/adapter.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/e2e_test/adapter.test.ts:169]: This is the real scenario that will happen in the SaaS, so was thinking that it's better to have a test for it

File: packages/netsuite-adapter/src/adapter_creator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/src/adapter_creator.ts:217]: I wrote another similar comment in the SaaS PR IIRC, we should think if this is the right direction as it has an implicit behavior that will cause some misalignment with the "installed versions"

File: packages/netsuite-adapter/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/README.md:32]: What TBH standa for?

[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/README.md:32]: TBA

File: packages/netsuite-adapter/src/adapter_creator.ts [hunk 6]
```diff
@@ -43,11 +55,19 @@ export const defaultCredentialsType = new ObjectType({
...
+      annotations: { message: 'SDF Token ID (for token-based auth)' },
     },
     tokenSecret: {
       refType: BuiltinTypes.STRING,
-      annotations: { message: 'SDF Token Secret' },
+      annotations: { message: 'SDF Token Secret (for token-based auth)' },
+    },
+    certificateId: {
+      refType: BuiltinTypes.STRING,
+      annotations: { message: 'SDF Certificate ID (for oauth)' },
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/adapter_creator.ts:68]: changed to OAuth 2.0
image

File: packages/netsuite-adapter/src/adapter_creator.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/adapter_creator.ts:217]: we can modify `AdapterSuccessInstallResult` :
```typescript
type AdapterSuccessInstallResult = {
  success: true
  installedVersion: string
  additionallyInstalledVersions?: string[]
}
```

WDYT?

[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/src/adapter_creator.ts]: Looks great!
Let's make sure that in the same deploy where this PR is bumped the SaaS uses `additionallyInstalledVersions`.

**Review by omrilit — APPROVED** [review]
👏

**Pull request merged**